posthog/api/person.py

```diff
@@ -613,11 +452,207 @@ def update(self, request, *args, **kwargs):
+    def calculate_funnel_persons(
+        self, request: request.Request
+    ) -> Dict[str, Tuple[List, Optional[str], Optional[str]]]:
+        if request.user.is_anonymous or not self.team:
+            return {"result": ([], None, None)}
+
+        filter = Filter(request=request, data={"insight": INSIGHT_FUNNELS}, team=self.team)
+        filter = prepare_actor_query_filter(filter)
+
+        funnel_actor_class = get_funnel_actor_class(filter)
+
+        actors, serialized_actors = funnel_actor_class(filter, self.team).get_actors()
+        initial_url = format_query_params_absolute_url(request, 0)
+        next_url = paginated_result(actors, request, filter.offset, filter.limit)
+
+        # cached_function expects a dict with the key result
+        return {"result": (serialized_actors, next_url, initial_url)}
+
+    @action(methods=["GET", "POST"], detail=False)
+    def path(self, request: request.Request, **kwargs) -> response.Response:
+        if request.user.is_anonymous or not self.team:
+            return response.Response(data=[])
+
+        results_package = self.calculate_path_persons(request)
+
+        if not results_package:
+            return response.Response(data=[])
+
+        people, next_url, initial_url = results_package["result"]
+
+        return response.Response(
+            data={
+                "results": [{"people": people, "count": len(people)}],
+                "next": next_url,
+                "initial": initial_url,
+                "is_cached": results_package.get("is_cached"),
+                "last_refresh": results_package.get("last_refresh"),
+            }
+        )
+
+    @cached_function
+    def calculate_path_persons(self, request: request.Request) -> Dict[str, Tuple[List, Optional[str], Optional[str]]]:
+        if request.user.is_anonymous or not self.team:
+            return {"result": ([], None, None)}
+
+        filter = PathFilter(request=request, data={"insight": INSIGHT_PATHS}, team=self.team)
+        filter = prepare_actor_query_filter(filter)
+
+        funnel_filter = None
+        funnel_filter_data = request.GET.get("funnel_filter") or request.data.get("funnel_filter")
+        if funnel_filter_data:
+            if isinstance(funnel_filter_data, str):
+                funnel_filter_data = json.loads(funnel_filter_data)
+            funnel_filter = Filter(data={"insight": INSIGHT_FUNNELS, **funnel_filter_data}, team=self.team)
+
+        people, serialized_actors = PathsActors(filter, self.team, funnel_filter=funnel_filter).get_actors()
+        next_url = paginated_result(serialized_actors, request, filter.offset, filter.limit)
+        initial_url = format_query_params_absolute_url(request, 0)
+
+        # cached_function expects a dict with the key result
+        return {"result": (serialized_actors, next_url, initial_url)}
+
+    @action(methods=["GET"], detail=False)
+    def trends(self, request: request.Request, *args: Any, **kwargs: Any) -> Response:
```

Comment:
without a `calculate_trend_persons` function, there is no caching here, which isn't great

(notice the @cached_function decorator in funnel/paths) That's the reason they're split into two functions.

Reply:
Good idea - I'll add that (maybe in a more reusable way...